data-connection-jdbc/src/main/java/io/micronaut/data/connection/jdbc/oracle/OracleClientInfoConnectionListener.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.connection.jdbc.oracle;
+
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.data.connection.ConnectionDefinition;
+import io.micronaut.data.connection.ConnectionStatus;
+import io.micronaut.data.connection.support.ConnectionClientInfoDetails;
+import io.micronaut.data.connection.support.ConnectionListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLClientInfoException;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A customizer for Oracle database connections that sets client information after opening and clears before closing.
+ *
+ * This customizer checks if the connection is an Oracle database connection and then sets the client information
+ * (client ID, module, and action) after opening the connection. It also clears these properties before closing the connection.
+ *
+ * @author radovanradic
+ * @since 4.10
+ */
+@EachBean(DataSource.class)
+@Requires(condition = OracleClientInfoCondition.class)
+@Internal
+final class OracleClientInfoConnectionListener implements ConnectionListener<Connection> {
+
+    /**
+     * Constant for the Oracle connection client info client ID property name.
+     */
+    private static final String ORACLE_CLIENT_ID = "OCSID.CLIENTID";
+    /**
+     * Constant for the Oracle connection client info module property name.
+     */
+    private static final String ORACLE_MODULE = "OCSID.MODULE";
+    /**
+     * Constant for the Oracle connection client info action property name.
+     */
+    private static final String ORACLE_ACTION = "OCSID.ACTION";
+    /**
+     * Constant for the Oracle connection database product name.
+     */
+    private static final String ORACLE_CONNECTION_DATABASE_PRODUCT_NAME = "Oracle";
+    private static final List<String> RESERVED_CLIENT_INFO_NAMES = List.of(ORACLE_CLIENT_ID, ORACLE_MODULE, ORACLE_ACTION);
+
+    private static final Logger LOG = LoggerFactory.getLogger(OracleClientInfoConnectionListener.class);
+
+    private final Map<Connection, Boolean> connectionSupportedMap = new ConcurrentHashMap<>(20);
+
+    @Override
+    public boolean supportsConnection(@NonNull ConnectionStatus<Connection> connectionStatus) {
+        return connectionSupportedMap.computeIfAbsent(connectionStatus.getConnection(), this::isOracleConnection);
```

Comment:
This creates a memory leak by collecting all the connections. After turning the connection definition to an annotation provider you should be able to check by simple `hasAnnotation`.

Inject the datasource and do unwrap + instanceof to disable the listener permanently.